### PR TITLE
extract changes

### PR DESCRIPTION
**Pull request by swaathee**
Changes to extract event

### CODE REVIEW

**Comment by toby-walker** [comment]
Add version change later

File: packages/clarity-js/src/data/encode.ts [hunk 1]
```diff
@@ -110,7 +110,7 @@ export default function(event: Event): void {
             let extractKeys = extract.keys;
             for (let e of extractKeys) {
                 tokens.push(e);
-                tokens.push(extract.data[e]);
+                tokens.push([].concat(...extract.data[e]));
```
Comment (toby-walker) [inline packages/clarity-js/src/data/encode.ts:113]: All you are doing here is making a shallow copy of the array into the tokens. Is that your intent?

File: packages/clarity-js/types/data.d.ts [hunk 2]
```diff
@@ -404,7 +406,7 @@ export interface UpgradeData {
 }
 
 export interface ExtractData {
-    [key: string]: string | number;
+    [key: number]: [string, string][];
```
Comment (toby-walker) [inline packages/clarity-js/types/data.d.ts:409]: What do the two element in the tuple (array) represent? Suggest you declare a type for the pair separately to make the intent clearer.

**Comment by toby-walker** [comment]
Add tests?

File: packages/clarity-js/src/data/upload.ts [hunk 3]
```diff
@@ -242,19 +243,26 @@ function delay(): number {
...
-            break;
-        case Constant.Upgrade:
-            // Upgrade current session to send back playback information
-            clarity.upgrade(Constant.Auto);
-            break;
-        case Constant.Action:
-            // Invoke action callback, if configured and has a valid value
-            if (config.action && parts.length > 1) { config.action(parts[1]); }
-            break;
+    let lines = payload && payload.length > 0 ? payload.split("\r\n") : [Constant.Empty];
```
Comment (toby-walker) [inline packages/clarity-js/src/data/upload.ts:246]: Isn't an empty list sufficient?

File: packages/clarity-js/src/data/extract.ts [hunk 4]
```diff
@@ -1,38 +1,35 @@
...
+}
+
+export function trigger(input: string): void { 
     try {
-        let e = config.extract;
-        if (!e) { return; }
-        for (let i = 0; i < e.length; i+=3) {
-            let source = e[i] as ExtractSource;
-            let key = e[i+1] as number;
+        var parts = input && input.length > 0 ? input.split(/ (.*)/) : [Constant.Empty];
```
Comment (toby-walker) [inline packages/clarity-js/src/data/extract.ts:18]: Wasn't this already split on earlier in the action handler parsing routine in upload.ts?

**Comment by toby-walker** [comment]
>         internal.log(Code.Config, Severity.Warning, e ? e.name : null);

Can you please add a bit of a comment about what this double indexed maps represent?

---
Refers to: packages/clarity-js/src/data/extract.ts:38 in f289dfc. [](commit_id = f289dfcaed4da6c6cfb19f0b4dd1d79bf3e1c0dd, deletion_comment = False)

**Review by toby-walker — APPROVED** [review]
:shipit:

File: packages/clarity-js/src/data/extract.ts [hunk 5]
```diff
@@ -49,13 +46,26 @@ export function clone(v: Syntax[]): Syntax[] {
...
 
-        for (let s in selectors) {
-            let node = document.querySelector(selectors[s] as string) as HTMLElement;
-            if (node) { update(s, node.innerText); }
+                let selectorData = selectors[key];
+                for (let selectorKey in selectorData) {
+                    let nodes = document.querySelectorAll(selectorData[selectorKey]) as NodeListOf<HTMLElement>;
+                    let text = [];
+                    if (nodes) {
+                        nodes.forEach((element) => {
```
Comment (sarveshnagpal) [inline packages/clarity-js/src/data/extract.ts:62]: Probably easier to use let text = node.map(e => e.innerText)

File: packages/clarity-decode/src/data.ts [hunk 6]
```diff
@@ -71,8 +71,19 @@ export function decode(tokens: Data.Token[]): DataEvent {
             let e = 2; // Start from 3rd index since first two are used for time & event
             let extract: Data.ExtractData = {};
             while (e < tokens.length) {
-                extract[tokens[e++] as string | number] = tokens[e++] as string;
+                if (typeof(tokens[e + 1]) == Constant.String) {
```
Comment (sarveshnagpal) [inline packages/clarity-decode/src/data.ts:74]: Assuming this is for backward compatibility. If so, could you please add a comment with version number so it's easy to track in future.

**Review by sarveshnagpal — APPROVED** [review]

File: packages/clarity-js/src/data/extract.ts (hunk 4 above)
Comment (swaathee) [reply packages/clarity-js/src/data/extract.ts:18]: this regex splits by the first space - so the parts are different.

File: packages/clarity-js/src/data/extract.ts [hunk 7]
```diff
@@ -49,13 +46,26 @@ export function clone(v: Syntax[]): Syntax[] {
...
+                            text.push(element.innerText);
```
[the remark this answers is not in the document]
Comment (swaathee) [reply packages/clarity-js/src/data/extract.ts:63]: done

File: packages/clarity-js/src/data/encode.ts (hunk 1 above)
Comment (swaathee) [reply packages/clarity-js/src/data/encode.ts:113]: we are converting the array of array into a single array

**Pull request merged**